flux-top: support ability to flip through queues via left/right arrow keys

chu11 opened the pull request:
It is inconvenient to have to restart flux-top to filter jobs by a different queue.

Support using the left/right arrow keys or h/l to "rotate" through all of the possible queues, including "all" queues.

Only minor thing is it's a little annoying that you don't know what queue (or "all" queues) you are currently rotating through if no jobs are in that queue.  So I added the queue name along with the fancy f in the upper left.  Could be somewhere else?  Originally I was going to add something right above the "QUEUE" header, until I realized the instance depth is already output there when you recurse down in `flux-top`.  We could also remove this, it's not super important.

This PR is built on top of #5032

grondo commented on t/t2801-top-cmd.t:
```diff
@@ -10,8 +10,6 @@ runpty="${SHARNESS_TEST_SRCDIR}/scripts/runpty.py"
 waitfile="${SHARNESS_TEST_SRCDIR}/scripts/waitfile.lua"
 testssh="${SHARNESS_TEST_SRCDIR}/scripts/tssh"
 
-export FLUX_URI_RESOLVE_LOCAL=t
```
This is fine, but might actually be more tractable long term to just fix the one test test that unsets the variable to run in a sub-shell, since every case of resolving a URI in the test should use `FLUX_URI_RESOLVE_LOCAL=t` (and it might not always be clear that a test will end up resolving a uri...)

chu11 replied:
good idea.  This change was from PR #5032 so I'll push a change in that PR.

grondo replied:
Oh, oops sorry!

grondo commented:
Nice! This is cool.

>  So I added the queue name along with the fancy f in the upper left. 

Was it also meant to be in yellow? It seems to give it greater importance than it should. What about just to the left of the version number, and specify `queue: <name>` like the other data in that row. It could also go in the blank like between resource summary and instance data line.

I also noticed there's a delay between switching the queue and getting an update in the resource summary.

chu11 commented:
> Was it also meant to be in yellow?

Oops, nope :P

> What about just to the left of the version number, and specify queue: <name> like the other data in that row. It could also go in the blank like between resource summary and instance data line.

Lemme see if I can sneak it in to the left of the version number, don't have a ton of space though.  Perhaps if queues are configured, there could just be an extra line below it.

> I also noticed there's a delay between switching the queue and getting an update in the resource summary.

Hmm yeah I notice it now too.  I draw the joblist pane and summary pane at the same time.  I'm wondering if the summary pane has to go get new queue specific data, then the update comes at the next heartbeat.  So gotta do something to make it update faster.

chu11 commented:
> Hmm yeah I notice it now too. I draw the joblist pane and summary pane at the same time. I'm wondering if the summary pane has to go get new queue specific data, then the summary pane re-draw comes at the next heartbeat. So gotta do something to make it re-draw earlier/immediately.

Ahhh my guess was almost correct.  the summary pane draw function doesn't do a query, so the query comes at the next heartbeat, which updates the summary stats.  Just gotta force the query before the draw.

```
diff --git a/src/cmd/top/keys.c b/src/cmd/top/keys.c
index 49e7fa0..5125d29 100644
--- a/src/cmd/top/keys.c
+++ b/src/cmd/top/keys.c
@@ -45,12 +45,14 @@ static void keys_cb (flux_reactor_t *r,
         case 'h':
         case KEY_LEFT:
             queues_prev (keys->top->queues);
+            summary_pane_query (keys->top->summary_pane);
             summary_pane_draw (keys->top->summary_pane);
             joblist_pane_draw (keys->top->joblist_pane);
             break;
         case 'l':
         case KEY_RIGHT:
             queues_next (keys->top->queues);
+            summary_pane_query (keys->top->summary_pane);
             summary_pane_draw (keys->top->summary_pane);
             joblist_pane_draw (keys->top->joblist_pane);
             break;
```

chu11 commented:
re-pushed, fixing the summary pane update per comment above.

after some trial and error, I liked adding the queue name on a new line in the summary pane

![image](https://user-images.githubusercontent.com/274859/229242403-19dfe55d-922c-45f6-b36f-cec19fcf53d0.png)

So we do lose a line ... and that's an empty line when there are no queues configured (or "all" queues are displayed), which doesn't look bad.  And we now have an extra line if we need to add anything in the future.

As an aside, I went back and forth between showing no queue name when all queues are listed vs listing something like "queue: all" ... any strong opinions on which way to go?  Right now it elects to show nothing when all queues are listed.

grondo commented:
Just an opinion, but not a fan of the extra line. Why can't `queue`  fit between `uptime` and `version` again?

My opinion is no need to put  `queue: all`. There may be a queue called "all"

chu11 commented:
> Just an opinion, but not a fan of the extra line. Why can't queue fit between uptime and version again?

It just seemed tight to me.  In the above we only have 18 spaces in between, so that's only a few spaces left and right of something like "queue: batch".

but now that I take a closer look "depth: %d" isn't very big and there's a lot of space between that and the "uptime" string.  What if I moved the uptime string a little to the left?

![image](https://user-images.githubusercontent.com/274859/229252820-f5afbc4d-9245-40cd-be6e-12e57316d352.png)

That doesn't look too bad, although "uptime" is a little left of center now if there's no queues.

chu11 commented:
I re-pushed, with the version that put the queue name on the same line as the uptime and version.

grondo commented:
Had another idea, the `queue: foo` entry is redundant whenever there are jobs to display. So how about just putting `No jobs in queue foo` in the job list window when there are no jobs in the selected queue. Then we don't have to clutter the bottom summary line with the queue that will be almost always unused or redundant.

chu11 commented:
> Had another idea, the queue: foo entry is redundant whenever there are jobs to display. So how about just putting No jobs in queue foo in the job list window when there are no jobs in the selected queue. Then we don't have to clutter the bottom summary line with the queue that will be almost always unused or redundant.

Oh that's a good idea.  I'll try to make it "pretty" by centering in in the joblist pane.

chu11 commented:
re-pushed

- I merged the commits from #5065 in here because it would have made things easier.
- per @grondo suggestion above, simply display "no jobs to display" in the joblist pane if there are none to display.  I do this even for any case, anonymous queue, per queue, etc.  I really like this, as it simply informs the user in an obvious way that nothing is broken, there's just nothing to display.

![image](https://user-images.githubusercontent.com/274859/232121784-1cf71d3b-e997-479b-90cb-d18a2b999bec.png)

grondo commented:
Nice, I like that better :+1:. However, I think the `no jobs to display` when not viewing a specific queue is an unnecessary change, though that's only my personal opinion. If you don't feel strongly about it, is there any way to suppress that message for the default case?

grondo commented:
>  I just liked it, b/c when you run flux top and nothing shows up, one might wonder why nothing shows up.

My point was just that this is how it has always been and since there's been no complaints, the change in behavior didn't seem necessary.

chu11 commented:
> My point was just that this is how it has always been and since there's been no complaints, the change in behavior didn't seem necessary.

That's a good point, that I am altering the original behavior.

chu11 commented:
re-pushed, per @grondo suggestion above, under anonymous queue or "all queues", do not display "no jobs to display"

grondo commented on src/cmd/top/joblist_pane.c:
```diff
@@ -99,6 +99,21 @@ void joblist_pane_draw (struct joblist_pane *joblist)
+        if (filter_queue)
+            mvwprintw (joblist->win,
+                       10,
+                       25,
```
Maybe just leave the note at the top? What if the terminal has < 25 rows?

chu11 replied:
yeah, i guess I wanted to center it to be pretty, but perhaps on line 2 or 3 would be preferable.

chu11 replied:
oh wait a second, 25 is the width ... it starts on line 10 ... do you feel that's too far down?

grondo replied:
Oh, oops. I felt I had a window that wasn't too small, and the text was near the bottom. But I forget how this works, is that line 10 of the pane? I'm fine with keeping it as is, it was just a suggestion.

chu11 replied:
Yeah, line 10 of the pane.  But we can bump it up a bit, maybe line 5 or something.

grondo reviewed (approved):
LGTM! A couple minor comments.

chu11 commented:
@grondo thanks.  Cleaned up that commit message and I moved up the message from line 10 to line 5 of the window.  It still looks pretty good on a quick test.
